### PR TITLE
8314891: Additional Zip64 extra header validation

### PR DESCRIPTION
**Pull request by LanceAndersen**
Please review this PR which improves the Zip64 extra header validation:

- Throw a ZipException If the extra len field is 0 and :
-- size, csize, or loc offset are set to 0xFFFFFFFF
-- disk starting number is set to 0xFFFF

- We have a valid size for the Zip64 extra header but we are missing the csize or loc fields if they are expected to be part of the header

Mach5 tiers 1-3 are clean

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314891](https://bugs.openjdk.org/browse/JDK-8314891): Additional Zip64 extra header validation (**Bug** - P4)

### Reviewers
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15650/head:pull/15650` \
`$ git checkout pull/15650`

Update a local copy of the PR: \
`$ git checkout pull/15650` \
`$ git pull https://git.openjdk.org/jdk.git pull/15650/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15650`

View PR using the GUI difftool: \
`$ git pr show -t 15650`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15650.diff">https://git.openjdk.org/jdk/pull/15650.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15650#issuecomment-1712527613)

### CODE REVIEW

File: test/jdk/java/util/zip/ZipFile/MissingZIP64EntriesTest.java [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/* @test
+ * @bug 8314891
+ * @summary Validate that a ZipException is thrown when the extra len is 0
+ * and the CEN size, csize,LOC offset fields are set to 0xFFFFFFFF, the disk
+ * starting number is set to 0xFFFF or when we have a valid Zip64 Extra header
+ * size but missing the corresponding field.
+ * @run junit MissingZIP64EntriesTest
```
Comment (mrserb) [inline test/jdk/java/util/zip/ZipFile/MissingZIP64EntriesTest.java:52]: Is this comment accurate? I think we should check 3 cases when the header extra len == 0, len == 8 and len ==16, but still do not contain all required information.

Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/MissingZIP64EntriesTest.java:52]: Clarified the comment to make it a bit clearer and also added additional tests

File: test/jdk/java/util/zip/ZipFile/MissingZIP64EntriesTest.java [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+     * Value to set the size, csize, or LOC offset CEN fields to when their
+     * actual value is stored in the Zip64 Extra Header
+     */
+    private static final long ZIP64_MAGICVAL = 0xFFFFFFFFL;
+
+    /**
+     * Value to set the Disk Start number offset CEN field to when the
+     * actual value is stored in the Zip64 Extra Header
+     */
+    private static final int  ZIP64_MAGICCOUNT = 0xFFFF;
```
[the remark this answers is not in the document]
Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/MissingZIP64EntriesTest.java]: Removed the extra space.  thank you for pointing it out

**Comment by eirbjo** [comment]
@LanceAndersen 

I noticed that this PR did not update `ZipInputStream.readLOC` to perform consistency validation between expected and actual extra field size and values. Any particular reason why processing of LOC headers was not made consistent with CEN?

**Comment by LanceAndersen** [comment]
> @LanceAndersen
> 
> I noticed that this PR did not update `ZipInputStream.readLOC` to perform consistency validation between expected and actual extra field size and values. Any particular reason why processing of LOC headers was not made consistent with CEN?

Intentional, as this was a follow on to the updates which were done previously to the CEN work in August, this is follow on cleanup.

Updates to ZipInputStream would be done separately under a separate PR or  could be done via your work on 8303866

**Comment by Foorcee** [comment]
Hey @LanceAndersen,

It was a common practice in obfuscation, to create zips with invalid headers. This change leads to a behavioral change that affects existing work processes.  Would it be possible to add an system property to restore the old behavior?

**Pull request closed without merging**